projects/concurrency/cache_test.go

```diff
@@ -0,0 +1,193 @@
+		c.Get("name")
+	})
+}
+
+func TestLastRecentlyUsedEntry(t *testing.T) {
+	c := make(map[string]cacheEntry[string, string])
+	c["1"] = cacheEntry[string, string]{value: "entry1", timeStamp: time.Now()}
+	c["2"] = cacheEntry[string, string]{value: "entry2", timeStamp: time.Now().Add(10 * time.Second)}
+	c["3"] = cacheEntry[string, string]{value: "entry3", timeStamp: time.Now().Add(20 * time.Second)}
+	c["4"] = cacheEntry[string, string]{value: "entry4", timeStamp: time.Now().Add(-30 * time.Second)}
+
+	key := getLastRecentlyUsedEntry(c)
+	require.Equal(t, "4", key)
+}
+
+func TestGetUnReadEntries(t *testing.T) {
+	tests := map[string]struct {
+		input    map[string]cacheEntry[string, string]
+		expected int
+	}{
+		"all entries read": {
+			input: map[string]cacheEntry[string, string]{
+				"name":     {value: "Anna", timeStamp: time.Now(), numberOfTimesRead: 2},
+				"hobby":    {value: "swimming", timeStamp: time.Now(), numberOfTimesRead: 1},
+				"lastName": {value: "Joe", timeStamp: time.Now(), numberOfTimesRead: 3},
+			},
+			expected: 0},
+		"no entries read": {
+			input: map[string]cacheEntry[string, string]{
+				"name":     {value: "Anna", timeStamp: time.Now(), numberOfTimesRead: 0},
+				"hobby":    {value: "swimming", timeStamp: time.Now(), numberOfTimesRead: 0},
+				"lastName": {value: "Joe", timeStamp: time.Now(), numberOfTimesRead: 0},
+			},
+			expected: 3},
+
+		"some entries read": {
+			input: map[string]cacheEntry[string, string]{
+				"name":     {value: "Anna", timeStamp: time.Now(), numberOfTimesRead: 2},
+				"hobby":    {value: "swimming", timeStamp: time.Now(), numberOfTimesRead: 0},
+				"lastName": {value: "Joe", timeStamp: time.Now(), numberOfTimesRead: 1},
+			},
+			expected: 1},
+	}
+
+	for key, value := range tests {
+		t.Run(key, func(t *testing.T) {
+			got := unReadEntriesInMap[string, string](value.input)
+			require.Equal(t, value.expected, got)
+		})
+	}
+}
+
+// linked list operations
+
+func TestInsertAtHeadOfList(t *testing.T) {
+	t.Run("insert at head", func(t *testing.T) {
+		list := &LinkedList[string]{}
+		list.insertAtHead(&Node[string]{key: "name"})
+		list.insertAtHead(&Node[string]{key: "hobby"})
+		list.insertAtHead(&Node[string]{key: "address"})
+		list.insertAtHead(&Node[string]{key: "role"})
+
+		expect := "role\naddress\nhobby\nname\n"
+
+		got := list.printList()
+
+		require.Equal(t, expect, got)
+		require.NotNil(t, list.tail)
+		require.Equal(t, "name", list.tail.key)
+	})
+}
+
+func TestMoveNodeToHead(t *testing.T) {
+	t.Run("empty list", func(t *testing.T) {
+		list_empty := &LinkedList[string]{}
+
+		list_empty.moveToHead(&Node[string]{key: "role"})
+		got := list_empty.printList()
+		require.Equal(t, "role\n", got)
+	})
+
+	t.Run("target node at tail", func(t *testing.T) {
+		list := &LinkedList[string]{}
+		node1 := &Node[string]{key: "node1"}
+		node2 := &Node[string]{key: "node2", prev: node1}
+		node3 := &Node[string]{key: "node3", prev: node2}
```

Comment:
Let's talk about a few different ways of testing here:
1. As you're doing here - setting up the complete internals of what you're testing
2. Setting up the minimum internals needed for something to work, i.e. _not_ always setting the `next` or `prev` pointers in each node if you know they're not needed for what you're testing
3. Using public interfaces (or possibly just-for-test interfaces)